Compare and write: add compare length limit, remove sector size limit

PepperJo opened the pull request:
**Change C++ API cmp&write semantics**

(unmodified) write bl semantic:
1. the write length is exactly "len"
1a. if "bl" is smaller than "len", return an error
1b. if "bl" is larger than "len", only write "len" bytes

modified compare semantics:
2. the compare length is capped by "len"
2a. if "cmp_bl" is smaller than "len", only compare that many bytes (keep for backwards compatibility, note that C API does only allow compare len == wire len)
2b. if "cmp_bl" is larger than "len", return an error (not be backwards compatible but it is a bug which should be fixed)

**Remove sector size limit**

Remove artificial sector size limit for compare and write. We can allow up to stripe unit bytes (if aligned, i.e. access is within stripe).

**Add tests**

Add general tests for compare and write for the C and C++ API and their different semantics.
Add tests for stripe unit bytes accesses.

PepperJo commented:
@idryomov

PepperJo commented:
Alternatively to return an error on cmp_bl.length() > len we also could cap it at len size, similar to the write bufferlist. But I do understand that the semantics might be weird then since for <= len cmp_bl.length() would be used but for >= len it is not which might be confusing...

idryomov commented:
> Alternatively to return an error on cmp_bl.length() > len we also could cap it at len size, similar to the write bufferlist. But I do understand that the semantics might be weird then since for <= len cmp_bl.length() would be used but for >= len it is not which might be confusing...

After thinking about this some more, I wonder if, rather than meticulously documenting and testing this `cmp_bl` direct pass-through quirk, we would be better off making the APIs behave the same, i.e. making `ofs` and `len` unconditionally define both compare and write steps.  The C++ API would then:
- allow both `cmp_bl` and `bl` to be larger than `len`, operating on first `len` bytes in that case
- immediately return `EINVAL` if either `cmp_bl` or `bl` is smaller than `len`

It would take a single sentence to document and would be consistent with other storage protocols ("Both Compare and Write must operate on the same LBA range" for NVMe, etc).  The only downside is that it would not be backwards compatible but:
- this quirk is not documented today
- the only known `compare_and_write` user uses the C API
- the C++ API is not widely used in general

so the odds of it breaking something in the wild are very low.

idryomov commented:
I know it was me who suggested the approach implemented in this PR -- backwards compatibility is always one of my top concerns -- but in this case we *are* breaking it anyway to fix the `cmp_bl` is larger than `len` case -- there is no way around it since librbd can't guarantee atomicity across stripe units.

What do you think?

idryomov commented on src/librbd/librbd.cc:
```diff
@@ -2674,7 +2674,7 @@ namespace librbd {
 
-    if (bl.length() < len) {
+    if (bl.length() < len || cmp_bl.length() > len) {
       tracepoint(librbd, write_exit, -EINVAL);
```
Unrelated, but please fold this in:
```suggestion
      tracepoint(librbd, compare_and_write_exit, -EINVAL);
```

PepperJo commented:
> After thinking about this some more, I wonder if, rather than meticulously documenting and testing this `cmp_bl` direct pass-through quirk, we would be better off making the APIs behave the same, i.e. making `ofs` and `len` unconditionally define both compare and write steps. The C++ API would then:
> 
>     * allow both `cmp_bl` and `bl` to be larger than `len`, operating on first `len` bytes in that case
> 
>     * immediately return `EINVAL` if either `cmp_bl` or `bl` is smaller than `len`
> 

I think it makes sense. Consistency is probably better here since it never was supposed to work this way. 
The only concern I have that if we change it like suggested, old code where `cmp_bl.length()` > `len` will still work but its
semantic would change. I'm not sure how that could be avoided.

idryomov commented:
> old code where `cmp_bl.length()` > `len` will still work but its
> semantic would change. I'm not sure how that could be avoided.

That code is already broken regardless because `cmp_bl.length()` is silently truncated to the object boundary on the OSD side, which means that it is highly variable based on `ofs`...

idryomov commented on src/test/librbd/test_librbd.cc:
```diff
@@ -2574,6 +2574,205 @@ TEST_F(TestLibRBD, TestDataPoolIO)
+
+  // Compare should fail because of mismatch
+  uint64_t mismatch_off;
+  written = rbd_compare_and_write(image, off, write_buffer.length(), mismatch_buffer.c_str(), write_buffer.c_str(), &mismatch_off, 0);
```
We generally try to adhere to 80 columns limit.

PepperJo commented:
@idryomov I now changed the behavior as discussed such that the C++ API now has the same semantics as the C API, i.e. "len" size is respected for both compare and write bufferlists passed. I changed `assemble_extent` accordingly. I also introduced new tests for the C and C++ API.

idryomov commented on src/librbd/io/ImageRequest.cc:
```diff
@@ -800,11 +804,9 @@ int ImageCompareAndWriteRequest<I>::prune_object_extents(
     return -EINVAL;
 
   I &image_ctx = this->m_image_ctx;
-  uint64_t sector_size = 512ULL;
```
It turns out that the sector size limit also protected us from RBD journal entries (`journaling` image feature) getting too large (the limit is the RBD object size - a couple dozen bytes).  Your test cases for a stripe unit sized compare and write assert as follows because with default striping stripe unit size is equal to object size:
```
[ RUN      ] TestLibRBD.TestCompareAndWrite
using new format!
../src/librbd/Journal.cc: In function 'uint64_t librbd::Journal<>::append_io_events(journal::EventType, const librbd::Journal::Bufferlists &, uint64_t, size_t, bool, int) [ImageCtxT = librbd::ImageCtx]' thread 7fc8660d7700 time 2022-05-03T10:18:48.635826+0000
../src/librbd/Journal.cc: 824: FAILED ceph_assert(bl.length() <= m_max_append_size)
 ceph version Development (no_version) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7fc8860b3b44]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x140798f) [0x7fc8860b398f]
 3: (librbd::Journal<librbd::ImageCtx>::append_io_events(librbd::journal::EventType, std::__cxx11::list<ceph::buffer::v15_2_0::list, std::allocator<ceph::buffer::v15_2_0::list> > const&, unsigned long, unsigned long, bool, int)+0x1a8) [0x4456da8]
 4: (librbd::Journal<librbd::ImageCtx>::append_io_event(librbd::journal::EventEntry&&, unsigned long, unsigned long, bool, int)+0x131) [0x44574a1]
 5: (librbd::io::ImageCompareAndWriteRequest<librbd::ImageCtx>::append_journal_event(bool)+0x152) [0x30c2652]
 6: (librbd::io::AbstractImageWriteRequest<librbd::ImageCtx>::send_request()+0x2a7) [0x30bf5e7]
 7: (librbd::io::ImageRequest<librbd::ImageCtx>::send()+0x2fc) [0x30bd7dc]
 8: (librbd::io::ImageRequest<librbd::ImageCtx>::aio_compare_and_write(librbd::ImageCtx*, librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&)+0x115) [0x30bde05]
 9: (librbd::io::ImageDispatch<librbd::ImageCtx>::compare_and_write(librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&, unsigned long, std::atomic<unsigned int>*, librbd::io::DispatchResult*, Context**, Context*)+0x2e5) [0x4646645]
[...]
```
You can reproduce by running the test binary with `RBD_FEATRES` environment variable set to `109`.

To address this, take a look at what is done for regular writes in `Journal::append_write_event` in `src/librbd/Journal.cc`.  We probably need to do the same for compare and write bufferlists instead of just forwarding to `Journal::append_io_event` in `ImageCompareAndWriteRequest::append_journal_event`.

idryomov replied:
It turns out that the sector size limit also protected us from RBD journal (`rbd feature enable <imagename> journaling`) entries getting too large (they are capped at object size - a couple dozen bytes).  Your test cases for a stripe unit sized compare and write trigger an assert because with default striping stripe unit size == object size:
```
[ RUN      ] TestLibRBD.TestCompareAndWrite
using new format!
../src/librbd/Journal.cc: In function 'uint64_t librbd::Journal<>::append_io_events(journal::EventType, const librbd::Journal::Bufferlists &, uint64_t, size_t, bool, int) [ImageCtxT = librbd::ImageCtx]' thread 7fc8660d7700 time 2022-05-03T10:18:48.635826+0000
../src/librbd/Journal.cc: 824: FAILED ceph_assert(bl.length() <= m_max_append_size)
 ceph version Development (no_version) quincy (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7fc8860b3b44]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x140798f) [0x7fc8860b398f]
 3: (librbd::Journal<librbd::ImageCtx>::append_io_events(librbd::journal::EventType, std::__cxx11::list<ceph::buffer::v15_2_0::list, std::allocator<ceph::buffer::v15_2_0::list> > const&, unsigned long, unsigned long, bool, int)+0x1a8) [0x4456da8]
 4: (librbd::Journal<librbd::ImageCtx>::append_io_event(librbd::journal::EventEntry&&, unsigned long, unsigned long, bool, int)+0x131) [0x44574a1]
 5: (librbd::io::ImageCompareAndWriteRequest<librbd::ImageCtx>::append_journal_event(bool)+0x152) [0x30c2652]
 6: (librbd::io::AbstractImageWriteRequest<librbd::ImageCtx>::send_request()+0x2a7) [0x30bf5e7]
 7: (librbd::io::ImageRequest<librbd::ImageCtx>::send()+0x2fc) [0x30bd7dc]
 8: (librbd::io::ImageRequest<librbd::ImageCtx>::aio_compare_and_write(librbd::ImageCtx*, librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&)+0x115) [0x30bde05]
 9: (librbd::io::ImageDispatch<librbd::ImageCtx>::compare_and_write(librbd::io::AioCompletion*, std::vector<std::pair<unsigned long, unsigned long>, std::allocator<std::pair<unsigned long, unsigned long> > >&&, ceph::buffer::v15_2_0::list&&, ceph::buffer::v15_2_0::list&&, unsigned long*, std::shared_ptr<neorados::IOContext>, int, ZTracer::Trace const&, unsigned long, std::atomic<unsigned int>*, librbd::io::DispatchResult*, Context**, Context*)+0x2e5) [0x4646645]
[...]
```
You can reproduce by invoking the test binary with `RBD_FEATURES` environment variable set to `109`.

To address this, take a look at what is done for regular writes in `Journal::append_write_event` in `src/librbd/Journal.cc`.  We probably need to do the same for compare and write bufferlists instead of just forwarding to `Journal::append_io_event` in `ImageCompareAndWriteRequest::append_journal_event`.

idryomov commented on src/test/librbd/test_librbd.cc:
```diff
@@ -2574,6 +2574,238 @@ TEST_F(TestLibRBD, TestDataPoolIO)
   rados_ioctx_destroy(ioctx);
 }
 
+TEST_F(TestLibRBD, TestCompareAndWrite)
```
Please split this into multiple tests, both by what is being tested and by sync vs async variant.  I see ten different tests here:

CompareAndWriteSuccess
AioCompareAndWriteSuccess
CompareAndWriteMismatch
AioCompareAndWriteMismatch
CompareAndWriteStripeUnitSuccess
AioCompareAndWriteStripeUnitSuccess
CompareAndWriteStripeUnitUnaligned
AioCompareAndWriteStripeUnitUnaligned
CompareAndWriteTooLarge
AioCompareAndWriteTooLarge

And similar for the C++ API, with `PP` suffix.

It will be somewhat repetitive, but that is fine.

idryomov replied:
Please split this into multiple test cases, both by what is being tested and by sync vs async variant.  I see ten different test cases here:

CompareAndWriteSuccess
AioCompareAndWriteSuccess
CompareAndWriteMismatch
AioCompareAndWriteMismatch
CompareAndWriteStripeUnitSuccess
AioCompareAndWriteStripeUnitSuccess
CompareAndWriteStripeUnitUnaligned
AioCompareAndWriteStripeUnitUnaligned
CompareAndWriteTooLarge
AioCompareAndWriteTooLarge

And similar for the C++ API, with `PP` suffix.

idryomov commented on src/test/librbd/test_librbd.cc:
```diff
@@ -2574,6 +2574,238 @@ TEST_F(TestLibRBD, TestDataPoolIO)
+  std::string cmp_buffer("This is a test");
+  std::string write_buffer("Write this !!!");
+  std::string mismatch_buffer("This will fail");
+  ASSERT_EQ(cmp_buffer.length(), write_buffer.length());
+  ASSERT_EQ(cmp_buffer.length(), mismatch_buffer.length());
```
Don't assert on things that are defined in the test itself.

idryomov replied:
Don't assert on something that is part of the test.

idryomov commented on src/test/librbd/test_librbd.cc:
```diff
@@ -2574,6 +2574,238 @@ TEST_F(TestLibRBD, TestDataPoolIO)
+  // large write test => we allow stripe unit size writes (aligned)
+  uint64_t stripe_unit;
+  rbd_get_stripe_unit(image, &stripe_unit);
+  // We require stripe_unit > cmp_bl
+  ASSERT_GT(stripe_unit, cmp_buffer.length());
+  // We require image size > 2 * stripe_unit
+  ASSERT_GT(size, stripe_unit * 2);
```
ditto

idryomov replied:
ditto

idryomov commented:
All my comments got duplicated by Github for some reason -- I cleanup up the duplicates, hopefully it looks OK on your end.

PepperJo commented:
> Github duplicated my comments for some reason -- I cleaned that up, hopefully it looks OK on your end.

No duplicates on my side. Thanks Ilya!

PepperJo commented on src/librbd/io/ImageRequest.cc:
```diff
@@ -760,12 +760,17 @@ uint64_t ImageCompareAndWriteRequest<I>::append_journal_event(
 void ImageCompareAndWriteRequest<I>::assemble_extent(
-  const LightweightObjectExtent &object_extent, bufferlist *bl) {
+  const LightweightObjectExtent &object_extent, bufferlist *bl,
+  bufferlist *cmp_bl) {
```
BTW is there a style checker/guide for the Ceph code base?

idryomov replied:
Please squash these auto-generated fixups into respective commits when you are done.

PepperJo replied:
Yes, will do. That was my intent anyway.

PepperJo commented:
@idryomov I squashed your changes, split the tests and changed the append to the journal similar to the write.

idryomov reviewed (changes requested):
Mostly nits -- I commented once for each but since there is a lot of duplication in the tests some apply for two (non-aio/aio) and some for four sites (non-aio/aio x c/cpp).

idryomov commented:
jenkins test make check

idryomov commented:
Also, the mock tests (`unittest_librbd` ninja target) need to be fixed up:
```
FAILED: src/test/librbd/CMakeFiles/unittest_librbd.dir/io/test_mock_ImageRequest.cc.o 
/usr/bin/ccache /usr/bin/clang++-13  -DBOOST_ALL_NO_LIB -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_DATE_TIME_DYN_LINK -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_RANDOM_DYN_LINK -DBOOST_SYSTEM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DHAVE_CONFIG_H -DTEST_LIBRBD_INTERNALS -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -Isrc/include -I../src -isystem /opt/ceph/include -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -isystem ../src/googletest/googlemock/include -isystem ../src/googletest/googlemock -isystem ../src/googletest/googletest/include -isystem ../src/googletest/googletest -isystem src/pmdk/src/include -g -Werror -fPIE   -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++17 -MD -MT src/test/librbd/CMakeFiles/unittest_librbd.dir/io/test_mock_ImageRequest.cc.o -MF src/test/librbd/CMakeFiles/unittest_librbd.dir/io/test_mock_ImageRequest.cc.o.d -o src/test/librbd/CMakeFiles/unittest_librbd.dir/io/test_mock_ImageRequest.cc.o -c ../src/test/librbd/io/test_mock_ImageRequest.cc
In file included from ../src/test/librbd/io/test_mock_ImageRequest.cc:52:
../src/librbd/io/ImageRequest.cc:751:28: error: no member named 'append_compare_and_write_event' in 'librbd::(anonymous namespace)::MockTestJournal'
  tid = image_ctx.journal->append_compare_and_write_event(extent.first,
        ~~~~~~~~~~~~~~~~~  ^
../src/test/librbd/io/test_mock_ImageRequest.cc:504:35: note: in instantiation of member function 'librbd::io::ImageCompareAndWriteRequest<librbd::(anonymous namespace)::MockTestImageCtx>::append_journal_event' requested here
  MockImageCompareAndWriteRequest mock_aio_image_write(
                                  ^
1 error generated.
```

idryomov reviewed (changes requested):
There are still a few unhandled comments.  While most of the them are nits, if you intended to ignore them, please add a comment stating that.  That said, I do want at least unused variables to be removed.

Otherwise LGTM!

PepperJo commented:
> There are still a few unhandled comments.
If that's the case that certainly was not my intend, sorry that I missed those. I will try to go over it once more and make sure all your comments are addressed.

idryomov commented:
> If that's the case that certainly was not my intent, sorry that I missed those. I will try to go over it once more and make sure all your comments are addressed.

No problem, Github tends to collapse things on the main tab so you may need to hit "Load more..." on "X hidden conversations" a bunch of times.  Currently there are 7 outstanding threads.

PepperJo commented on src/test/librbd/test_librbd.cc:
```diff
@@ -2574,6 +2574,519 @@ TEST_F(TestLibRBD, TestDataPoolIO)
+  ASSERT_EQ(5U, mismatch_off);
+
+  // check nothing was written
+  ssize_t read = rbd_read(image, off, read_buffer.length(), &read_buffer[0]);
```
I changed it for the cases where functions where called like this but I would keep &..[X] for these (if it makes sense to you) because I think it is easier to read:
`std::string cmp_buffer("This is a test");
    ceph::bufferlist cmp_bl;
    cmp_bl.append(&cmp_buffer[0], 5);
    cmp_bl.append(&cmp_buffer[5], 3);
    cmp_bl.append(&cmp_buffer[8], 2);
    cmp_bl.append(&cmp_buffer[10], 4);`

idryomov replied:
Sure, makes sense.

PepperJo commented on src/test/librbd/test_librbd.cc:
```diff
@@ -3047,6 +3560,1057 @@ TEST_F(TestLibRBD, TestIOPP)
+                                      small_bl, /* cmp_bl */
+                                      write_bl,
+                                      &mismatch_off, 0);
+    ASSERT_EQ(-EINVAL, written);
```
It makes sense I added `ASSERT_EQ(0U, mismatch_off);` everywhere even if it failed "early"

idryomov commented:
> I changed it for the cases where functions [...]

Did you mean to push?  I see that you resolved all but that conversation but no code update.

PepperJo commented:
Not yet, I just rebased to latest master and cleaning up my commits I plan to push soon.

idryomov commented:
jenkins test make check

idryomov commented:
Added release notes and addressed a couple of still remaining `mismatch_off` nits.

idryomov commented:
jenkins test api

idryomov commented:
jenkins test make check

PepperJo commented:
Thanks Ilya!

idryomov commented:
jenkins test api

idryomov commented:
More fixes for issues turned up by teuthology.  @CongMinYin please see the PWL adjustment -- I already tested it but it would be good if you double checked it.  I'm also thinking of actually fixing `discard_set` to not populate the extent with zeroes but that is for another PR.

idryomov commented:
jenkins test api

idryomov commented:
jenkins test api

CongMinYin commented:
> More fixes for issues turned up by teuthology. @CongMinYin please see the PWL adjustment -- I already tested it but it would be good if you double checked it. I'm also thinking of actually fixing `discard_set` to not populate the extent with zeroes but that is for another PR.

LGTM.  thanks!

idryomov commented:
No actual failures:

https://pulpito.ceph.com/dis-2022-08-14_06:51:50-rbd-wip-dis-testing-2-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-08-14_17:49:24-rbd-wip-dis-testing-2-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-08-14_20:17:19-rbd-wip-dis-testing-2-distro-default-smithi/
